snapcraft/_store.py

```diff
@@ -828,16 +824,98 @@ def collaborate(snap_name, key):
+
     # There is a possibility that the `authority-id` to be `canonical`,
     # which should be changed to the `publisher_id` to match the signing key.
     assertion['authority-id'] = publisher_id
 
-    assertion = _sign_assertion(snap_name, assertion, key, 'developers')
+    signed_assertion = _sign_assertion(snap_name, assertion, key, 'developers')
+    store.push_assertion(snap_id, signed_assertion, 'developers')
+
 
-    store.push_assertion(snap_id, assertion, 'developers')
+_COLLABORATION_HEADER = dedent("""\
+    # Change which developers may build or upload snaps on the publisher's behalf.
+    #
+    # Sample entry:
+    #
+    # developers:
+    #   - developer-id: "dev-one"      # Which developer
+    #     since: "2017-02-10 08:35:00" # When contributions started
+    #     until: "2018-02-10 08:35:00" # When contributions ceased (optional)
+    #
+    # All timestamps are UTC, and the "now" special string will be replaced by
+    # the current time. Do not remove entries or use an until time in the past
+    # unless you want existing snaps provided by the developer to stop working.""") # noqa
+
+
+def _edit_collaborators(developers):
+    """Spawn an editor to modify the snap-developer assertion for a snap."""
+    editor_cmd = os.getenv('EDITOR', 'vi')
+
+    developer_wrapper = {'developers': _reformat_time_from_assertion(
+        developers)}
+
+    with tempfile.NamedTemporaryFile() as ft:
+        ft.close()
+        with open(ft.name, 'w') as fw:
+            print(_COLLABORATION_HEADER, file=fw)
+            yaml.dump(developer_wrapper, stream=fw, default_flow_style=False)
+        subprocess.check_call([editor_cmd, ft.name])
+        with open(ft.name, 'r') as fr:
+            developers = yaml.load(fr).get('developers')
+    return _reformat_time_for_assertion(developers)
+
+
+def _reformat_time_from_assertion(developers):
+    reformatted_developers = []
+    for developer in developers:
+        developer_it = {'developer-id': developer['developer-id']}
+        for range in ['since', 'until']:
+            if range in developer:
+                date = datetime.strptime(developer[range],
+                                         '%Y-%m-%dT%H:%M:%S.%fZ')
+                developer_it[range] = datetime.strftime(date,
+                                                        '%Y-%m-%d %H:%M:%S')
+        reformatted_developers.append(developer_it)
+    return reformatted_developers
+
+
+def _reformat_time_for_assertion(developers):
+    reformatted_developers = []
+    for developer in developers:
+        developer_it = {'developer-id': developer['developer-id']}
+        for range_ in ['since', 'until']:
+            if range_ in developer:
+                if developer[range_] == 'now':
+                    date = datetime.now()
+                else:
+                    date = datetime.strptime(developer[range_],
+                                             '%Y-%m-%d %H:%M:%S')
```

Comment:
Was wondering if we could default 00:00:00 when there is no corresponding HH:MM:SS input, to provide easy UE, but that not a huge issue.